core/src/main/java/nu/yona/server/goals/service/GoalService.java

```diff
@@ -85,23 +87,43 @@ public GoalDTO addGoal(UUID userID, GoalDTO goal, Optional<String> message)
 	@Transactional
 	public GoalDTO updateGoal(UUID userID, UUID goalID, GoalDTO newGoalDTO, Optional<String> message)
 	{
-		newGoalDTO.validate();
 		User userEntity = userService.getUserByID(userID);
 		Goal existingGoal = getGoalEntity(userEntity, goalID);
+
+		verifyGoalUpdate(existingGoal, newGoalDTO);
+
+		if (newGoalDTO.isGoalChanged(existingGoal))
+		{
+			updateGoal(userEntity, existingGoal, newGoalDTO, message);
+		}
+
+		return GoalDTO.createInstance(existingGoal);
+	}
+
+	private void verifyGoalUpdate(Goal existingGoal, GoalDTO newGoalDTO)
+	{
+		newGoalDTO.validate();
 		GoalDTO existingGoalDTO = GoalDTO.createInstance(existingGoal);
 		assertNoTypeChange(newGoalDTO, existingGoalDTO);
 		assertNoActivityCategoryChange(newGoalDTO, existingGoalDTO);
+	}
 
+	private void updateGoal(User userEntity, Goal existingGoal, GoalDTO newGoalDTO, Optional<String> message)
+	{
 		UserAnonymized userAnonymizedEntity = userEntity.getAnonymized();
-
-		newGoalDTO.updateGoalEntity(existingGoal); // TODO: Change the implementation to retain the old goal, linked to related
-													// activities. With that, also check whether the goal really changed.
-		userAnonymizedEntity.addGoal(existingGoal);
+		cloneExistingGoalAsHistoryItem(userAnonymizedEntity, existingGoal);
+		newGoalDTO.updateGoalEntity(existingGoal);
 		userAnonymizedService.updateUserAnonymized(userAnonymizedEntity.getID(), userAnonymizedEntity);
 
 		broadcastGoalChangeMessage(userEntity, existingGoal, GoalChangeMessage.Change.GOAL_CHANGED, message);
+	}
 
-		return GoalDTO.createInstance(existingGoal);
+	private void cloneExistingGoalAsHistoryItem(UserAnonymized userAnonymizedEntity, Goal existingGoal)
+	{
+		Goal historyGoal = existingGoal.cloneAsHistoryItem();
+		existingGoal.setPreviousGoal(historyGoal);
+		WeekActivity.getRepository().findByGoal(existingGoal).stream().forEach(a -> a.setGoal(historyGoal));
```

Comment:
I hope this will perform. Should we not write a specific update query in the repositories to make it more efficient (`update DayActivity set goal = :historyGoal where goal = :goal`)?